skafka/src/main/scala/com/evolutiongaming/skafka/consumer/RebalanceCallback.scala

```diff
@@ -0,0 +1,333 @@
+          }
+        case c: WithConsumer[A1] =>
+          Try { c.f(consumer) } match {
+            case Success(a) =>
+              ss match {
+                case Nil     => a.pure[Try]
+                case s :: ss => loop(s(a), ss)
+              }
+            case Failure(a) => a.raiseError[Try, A1]
+          }
+        case c: Bind[F, _, A1] =>
+          loop(c.source, c.fs.asInstanceOf[S] :: ss)
+        case Error(a) =>
+          a.raiseError[Try, A1]
+      }
+    }
+
+    for {
+      a <- loop(rebalanceCallback, List.empty)
+      a <- Try { a.asInstanceOf[A] }
+    } yield a
+  }
+
+  private final case class Pure[+A](a: A) extends RebalanceCallback[Nothing, A]
+
+  private final case class Bind[F[_], S, +A](source: RebalanceCallback[F, S], fs: S => RebalanceCallback[F, A])
+      extends RebalanceCallback[F, A]
+
+  private final case class Lift[F[_], A](fa: F[A]) extends RebalanceCallback[F, A]
+
+  private final case class WithConsumer[+A](f: ConsumerJ[_, _] => A) extends RebalanceCallback[Nothing, A]
+
+  private final case class Error(throwable: Throwable) extends RebalanceCallback[Nothing, Nothing]
+
+  implicit class RebalanceCallbackOps[F[_], A](val self: RebalanceCallback[F, A]) extends AnyVal {
+
+    def map[B](f: A => B): RebalanceCallback[F, B] = flatMap(a => pure(f(a)))
+
+    def flatMap[B](f: A => RebalanceCallback[F, B]): RebalanceCallback[F, B] = Bind(self, f)
+
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, A] = {
+      self match {
+        case Pure(a)          => Pure(a)
+        case Bind(source, f)  => Bind(source.mapK(fg), f andThen (_.mapK(fg)))
+        case Lift(fa)         => Lift(fg(fa))
+        case WithConsumer(f)  => WithConsumer(f)
+        case Error(throwable) => Error(throwable)
+      }
+    }
+
+  }
+
+  implicit class RebalanceCallbackNothingOps[A](val self: RebalanceCallback[Nothing, A]) extends AnyVal {
+    def effectAs[F[_]]: RebalanceCallback[F, A] = self
+    def apply[F[_]]: RebalanceCallback[F, A]    = effectAs
+  }
+
+  private[consumer] object Helpers {
+
+    def committed1(
+      f: ConsumerJ[_, _] => MapJ[TopicPartitionJ, OffsetAndMetadataJ]
+    ): RebalanceCallback[Nothing, Map[TopicPartition, OffsetAndMetadata]] = {
+      for {
+        result <- WithConsumer(f)
+        result <- fromTry(committedOffsetsF[Try](result))
+      } yield result
+    }
+
+    def offsets1(
+      f: ConsumerJ[_, _] => MapJ[TopicPartitionJ, LongJ]
+    ): RebalanceCallback[Nothing, Map[TopicPartition, Offset]] = {
+      for {
+        result <- WithConsumer(f)
+        result <- fromTry(offsetsMapF[Try](result))
+      } yield result
+    }
+
+  }
+
+}
+
+abstract private[consumer] class RebalanceCallbackInstances {
+  implicit val catsMonadForRebalanceCallbackWithNoEffect: StackSafeMonad[RebalanceCallback[Nothing, *]] =
+    catsMonadForRebalanceCallback[Nothing]
+
+  implicit def catsMonadForRebalanceCallback[F[_]]: StackSafeMonad[RebalanceCallback[F, *]] =
```

Comment:
please use `Monad[RebalanceCallback[F, *]]` as return type here

Reply:
fixed in e3ac3b60